pkg/server/image_pull.go

```diff
@@ -182,216 +166,3 @@ func ParseAuth(auth *runtime.AuthConfig) (string, string, error) {
-	if err != nil {
-		return "", "", "", fmt.Errorf("failed to get fetcher for ref %q: %v", ref, err)
-	}
-	// Currently, the resolved image name is the same with ref in docker resolver,
-	// but they may be different in the future.
-	// TODO(random-liu): Always resolve image reference and use resolved image name in
-	// the system.
-
-	glog.V(4).Infof("Start downloading resources for image %q", ref)
-	resources := newResourceSet()
-	resourceTrackHandler := containerdimages.HandlerFunc(func(ctx gocontext.Context, desc imagespec.Descriptor) (
-		[]imagespec.Descriptor, error) {
-		resources.add(remotes.MakeRefKey(ctx, desc))
-		return nil, nil
-	})
-	// Fetch all image resources into content store.
-	// Dispatch a handler which will run a sequence of handlers to:
-	// 1) track all resources associated using a customized handler;
-	// 2) fetch the object using a FetchHandler;
-	// 3) recurse through any sub-layers via a ChildrenHandler.
-	// Support schema1 image.
-	var (
-		schema1Converter *schema1.Converter
-		handler          containerdimages.Handler
-	)
-	if desc.MediaType == containerdimages.MediaTypeDockerSchema1Manifest {
-		schema1Converter = schema1.NewConverter(c.contentStoreService, fetcher)
-		handler = containerdimages.Handlers(
-			resourceTrackHandler,
-			schema1Converter,
-		)
-	} else {
-		handler = containerdimages.Handlers(
-			resourceTrackHandler,
-			remotes.FetchHandler(c.contentStoreService, fetcher),
-			containerdimages.ChildrenHandler(c.contentStoreService),
-		)
-	}
-	if err := containerdimages.Dispatch(ctx, handler, desc); err != nil {
-		// Dispatch returns error when requested resources are locked.
-		// In that case, we should start waiting and checking the pulling
-		// progress.
-		// TODO(random-liu): Check specific resource locked error type.
-		glog.V(5).Infof("Dispatch for %q returns error: %v", ref, err)
-	}
-	// Wait for the image pulling to finish
-	if err := c.waitForResourcesDownloading(ctx, resources.all()); err != nil {
-		return "", "", "", fmt.Errorf("failed to wait for image %q downloading: %v", ref, err)
-	}
-	glog.V(4).Infof("Finished downloading resources for image %q", ref)
-	if schema1Converter != nil {
-		desc, err = schema1Converter.Convert(ctx)
-		if err != nil {
-			return "", "", "", fmt.Errorf("failed to convert schema 1 image %q: %v", ref, err)
-		}
-	}
-
-	// In the future, containerd will rely on the information in the image store to perform image
-	// garbage collection.
-	// For now, we simply use it to store and retrieve information required for pulling an image.
-	// @stevvooe said we should `Put` before downloading content, However:
-	// 1) Containerd client put image metadata after downloading;
-	// 2) We need desc returned by schema1 converter.
-	// So just put the image metadata after downloading now.
-	// TODO(random-liu): Fix the potential garbage collection race.
-	repoDigest, repoTag := getRepoDigestAndTag(namedRef, desc.Digest, schema1Converter != nil)
```

Comment:
Currently, our internal image store is maintaining the Image ID -> RepoDigests/RepoTags mapping, which is for the requirement of CRI.

However, when user create a container or pulling a image, they usually give us RepoDigest/RepoTag. So we have to do the RepoDigest/RepoTag -> Image ID (and other Information) conversion, which we are currently leveraging containerd image metadata store to do that. That's why we should put all `RepoDigest/RepoTag` into the metadata store.

As for why we also need to put image ID - I was told that containerd will do image gc based on reference, an entry in image metadata store is a reference to the corresponding snapshot and content. Other references are not unique enough to identify an image, e.g. repo tag could be reused by another image; schema 1 image doesn't have a repo digest etc.; image id seems to be the only thing we can use the reference the image and keep it from being garbage collected.
I think it's because other references are not unique to identify